### PR TITLE
8260043: Reduce allocation in sun.net.www.protocol.jar.Handler.parseURL

### PR DESCRIPTION
**Pull request by eirbjo**
By moving string splitting and concatenation into the canonizeString utility, we can defer allocation until we determine that canonization is required. This saves two string allocations and a string concat for the common case where canonization is not required.

As a refactoring, move ParseUtil.canonizeString/doCanonize into Handler since that's the only call site.

Finally, let's rename the method to canonizalizeString, since canonization is an rather unrelated term. 

/issue 8260043

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8260043](https://bugs.openjdk.java.net/browse/JDK-8260043): Reduce allocation in sun.net.www.protocol.jar.Handler.parseURL

### Reviewers
 * [Claes Redestad](https://openjdk.java.net/census#redestad) (@cl4es - **Reviewer**)
 * [Chris Hegarty](https://openjdk.java.net/census#chegar) (@ChrisHegarty - **Reviewer**) ⚠️ Review applies to f4b543d6f1765df5032b2dba950f04962cf55ab5

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/2167/head:pull/2167`
`$ git checkout pull/2167`

### CODE REVIEW

File: src/java.base/share/classes/sun/net/www/protocol/jar/Handler.java [hunk 1]
```diff
@@ -216,4 +212,51 @@ private String parseContextSpec(URL url, String spec) {
...
+            return file;
+        } else {
+            // Defer substring and concat until canonicalization is required
+            String before = file.substring(0, off);
+            String after = file.substring(off);
+            return before + doCanonize(after);
+        }
+    }
+
+    private static String doCanonize(String file) {
```
Comment (cl4es) [inline src/java.base/share/classes/sun/net/www/protocol/jar/Handler.java:232]: nit: doCanonicalize

**Review by cl4es — APPROVED** [review]
LGTM - just a small nit that you seem to have forgotten to change canonize to canonicalize in some places.

File: src/java.base/share/classes/sun/net/www/protocol/jar/Handler.java (hunk 1 above)
Comment (eirbjo) [reply src/java.base/share/classes/sun/net/www/protocol/jar/Handler.java]: Good catch! Not enough coffee :-)

**Comment by eirbjo** [comment]
Newbie question: Is a copyright year updates required when code is simply moved around like this?

More specifically, is it required when strictly deleting code from a file, like ParseUtil in this case?

**Review by ChrisHegarty — APPROVED** [review]
Looks good to me.

File: src/java.base/share/classes/sun/net/www/protocol/jar/Handler.java [hunk 2]
```diff
@@ -166,10 +165,7 @@ protected void parseURL(URL url, String spec,
 
             // Canonize the result after the bangslash
```
Comment (AlanBateman) [inline src/java.base/share/classes/sun/net/www/protocol/jar/Handler.java:166]: Maybe we can also replace "Canonize" in the comment while we are in the area.

**Comment by ChrisHegarty** [comment]
> Newbie question: Is a copyright year updates required when code is simply moved around like this?
> 
> More specifically, is it required when strictly deleting code from a file, like ParseUtil in this case?

I would prefer if the copyright year range was updated in these files - this is a non-trivial change.

**Review by cl4es — APPROVED** [review]
Looks good. (You need to cast /integrate again for us to be able to /sponsor)

**Pull request closed without merging**